aziot/src/init.rs

```diff
@@ -0,0 +1,867 @@
+	const HOSTNAME_KEY: &str = "HostName";
+	const DEVICEID_KEY: &str = "DeviceId";
+	const SHAREDACCESSKEY_KEY: &str = "SharedAccessKey";
+
+	let mut iothub_hostname = None;
+	let mut device_id = None;
+	let mut symmetric_key = None;
+
+	for sections in connection_string.split(';') {
+		let mut parts = sections.split('=');
+		match parts.next().expect("str::split always returns at least one part") {
+			HOSTNAME_KEY => iothub_hostname = parts.next(),
+			DEVICEID_KEY => device_id = parts.next(),
+			SHAREDACCESSKEY_KEY => symmetric_key = parts.next(),
+			_ => (), // Ignore extraneous component in the connection string
+		}
+	}
+
+	let iothub_hostname = iothub_hostname.ok_or(r#"required parameter "HostName" is missing"#)?;
+
+	let device_id = device_id.ok_or(r#"required parameter "DeviceId" is missing"#)?;
+
+	let symmetric_key = symmetric_key.ok_or(r#"required parameter "SharedAccessKey" is missing"#)?;
+	let symmetric_key =
+		base64::decode(symmetric_key)
+		.map_err(|err| format!(r#"connection string's "SharedAccessKey" parameter could not be decoded from base64: {}"#, err))?;
+
+	Ok((iothub_hostname.to_owned(), device_id.to_owned(), symmetric_key))
+}
+
+fn convert_path_or_uri_to_uri(value: String) -> Option<String> {
+	if value.starts_with("file://") || value.starts_with("pkcs11:") {
+		Some(value)
+	}
+	else {
+		match url::Url::from_file_path(&value) {
+			Ok(value) => Some(value.to_string()),
+			Err(()) => {
+				println!("Could not convert path to a file:// URI");
+				None
+			},
+		}
+	}
+}
+
+fn get_hostname() -> Result<String, crate::Error> {
+	if cfg!(test) {
+		Ok("my-device".to_owned())
+	}
+	else {
+		let mut hostname = vec![0_u8; 256];
+		let hostname = nix::unistd::gethostname(&mut hostname).map_err(|err| format!("could not get machine hostname: {}", err))?;
+		let hostname = hostname.to_str().map_err(|err| format!("could not get machine hostname: {}", err))?;
+		Ok(hostname.to_owned())
+	}
+}
+
+/// Prompts the user to enter an EST server URL, and fixes it to have a "/.well-known/est" component if it doesn't already.
+fn read_est_url(stdin: &mut impl std::io::BufRead, prompt_question: &str) -> Result<url::Url, crate::Error> {
+	loop {
+		let url = prompt(stdin, prompt_question)?;
+		let url =
+			if url.contains("/.well-known/est") {
+				url
+			}
+			else {
+				format!("{}/.well-known/est", url)
+			};
+		let url = url.trim_end_matches("/simpleenroll");
+		match url.parse::<url::Url>() {
+			Ok(url) if url.scheme() == "http" || url.scheme() == "https" => break Ok(url),
+			Ok(_) => println!("Could not parse response as an http or https URL"),
+			Err(err) => println!("Could not parse response as an http or https URL: {}", err),
+		}
+	}
+}
+
+fn create_dir_all(
+	path: &(impl AsRef<std::path::Path> + ?Sized),
+	user: &nix::unistd::User,
+	mode: u32,
+) -> Result<(), crate::Error> {
+	let path = path.as_ref();
+	let path_displayable = path.display();
+
+	let () =
```

Comment:
any reason for these `let () =` bindings?

Reply:
https://github.com/Azure/iot-identity-service/blob/1499fb72dbc2ffc89ef5a78e474db50f17672a8c/CONTRIBUTING.md See the note for `clippy::let_unit_value`

Reply:
Huh, I can't say I've ever seen that pattern before 🤔

While I don't think it's something I'd do myself, I can see how it might make it easier for users unfamiliar with the specific function understand that the return value is unit.